skfem/mesh/mesh.py

```diff
@@ -757,3 +731,60 @@ def element_finder(self, mapping=None):
 
         """
         raise NotImplementedError
+
+    @classmethod
+    def periodic(cls, mesh, ix, ix0):
+        """Initialize a periodic mesh from a standard (nonperiodic) mesh.
+
+        The mesh types that can describe a discontinuous topology include:
+
+        - :class:`~skfem.mesh.MeshTri1DG`
+        - :class:`~skfem.mesh.MeshQuad1DG`
+
+        Parameters
+        ----------
+        mesh
+            The mesh to periodify.
+        ix
+            An array of nodes to eliminate.  They are replaced by the nodes in
+            the array ``ix``.
+        ix0
+            An array of nodes left unchanged.
+
+        """
+        assert cls.elem.interior_dofs > 0
+        assert cls.elem.refdom == mesh.elem.refdom
+        assert len(ix) == len(ix0)
+
+        # reorder vertices: eliminated nodes must have highest index values
+        nix = np.arange(mesh.nvertices - len(ix),
+                        mesh.nvertices,
+                        dtype=np.int64)
+        remap = np.zeros(mesh.nvertices, dtype=np.int64)
+        remap[ix] = nix
+        oix = (remap == 0).nonzero()[0]
+        remap[oix] = np.arange(mesh.nvertices - len(ix), dtype=np.int64)
+
+        doflocs = np.hstack((mesh.doflocs[:, oix], mesh.doflocs[:, ix]))
+        t = remap[mesh.t]
+
+        reordered_mesh = replace(
+            mesh,
+            doflocs=doflocs,
+            t=t,
+            sort_t=False,
+        )
+
+        # make periodic
+        reremap = np.arange(mesh.nvertices, dtype=np.int64)
+        ix1 = remap[ix]
+        ix2 = remap[ix0]
+        reremap[ix1] = ix2
+
+        periodic_mesh = cls.from_mesh(reordered_mesh, reremap[t])
+
+        # store reordered mesh and reverse mapping
+        periodic_mesh._unperiodic = reordered_mesh
+        periodic_mesh._nodes = reremap
+
+        return periodic_mesh
```

Comment:
Should this not be part of `Mesh` but instead some `MeshDG` which is inherited by all of the new DG mesh types?

Reply:
Right, yes, ...  I'm no OO-theorist, but I have idly wondered (since before ) whether `skfem.Mesh` might benefit from multiple inheritance; e.g. whether `MeshTri1` might derive from both `Mesh2D` and `MeshSimplicial` (and to what extent this might indeed define it, as it kind of does mathematically, I think).  I don't know whether that makes sense in practice or offers any practical advantage, but it's something that I daydreamed about.

And so yes, would that make sense here?  It makes sense to me, I think, if it works and doesn't incur too much overhead.

What are the other ways of implementing that kind of thing?  `functools.singledispatch` as in #287, [`abc`](https://docs.python.org/3/library/abc.html#module-abc) as suggested as an alternative [at the time](https://github.com/kinnala/scikit-fem/pull/287#pullrequestreview-336936688)?  

Yeah, looking at `mesh.py` now, it doesn't look too good having the `.periodic` method there when it only applies to the discontinuous subclasses and then they probably do still want to derive from their dimensional superclasses via their cell-type superclasses, e.g. `MeshTri1DG` <- `MeshTri1` <- `Mesh2D` <- `Mesh`.  But then nor would one wish to have to duplicate the implementation of periodicity in each discontinuous subclass.

Is there any drawback to multiple inheritance here?  I might look to see where this is tested and try it out.

Reply:
I read about dataclasses and multiple inheritance. It seems to be a messy subject because of `__post_init__` but I'll try to make a proposition and update this PR.